EKF/mag_control.cpp

```diff
@@ -0,0 +1,355 @@
+
+	if (_control_status.flags.mag_3D
+	    && _control_status_prev.flags.update_mag_states_only
+	    && !_control_status.flags.update_mag_states_only) {
+		// When re-commencing use of magnetometer to correct vehicle states
+		// set the field state variance to the observation variance and zero
+		// the covariance terms to allow the field states re-learn rapidly
+		clearMagCov();
+
+		for (uint8_t index = 0; index <= 5; index ++) {
+			P[index + 16][index + 16] = sq(_params.mag_noise);
+		}
+
+		// save covariance data for re-use when auto-switching between heading and 3-axis fusion
+		saveMagCovData();
+	}
+}
+
+void Ekf::checkMagDeclRequired()
+{
+	// if we are using 3-axis magnetometer fusion, but without external NE aiding,
+	// then the declination must be fused as an observation to prevent long term heading drift
+	// fusing declination when gps aiding is available is optional, but recommended to prevent
+	// problem if the vehicle is static for extended periods of time
+	const bool user_selected = (_params.mag_declination_source & MASK_FUSE_DECL);
+	const bool not_using_ne_aiding = !_control_status.flags.gps;
+	_control_status.flags.mag_dec = (_control_status.flags.mag_3D && (not_using_ne_aiding || user_selected));
+}
+
+void Ekf::checkMagInhibition()
+{
+	_mag_use_inhibit = shouldInhibitMag();
+	if (!_mag_use_inhibit) { 
+		_mag_use_not_inhibit_us = _imu_sample_delayed.time_us;
+	}
+
+	// If magnetometer use has been inhibited continuously then a yaw reset is required for a valid heading
+	if (uint32_t(_imu_sample_delayed.time_us - _mag_use_not_inhibit_us) > (uint32_t)5e6) {
+		_mag_inhibit_yaw_reset_req = true;
+	}
+}
+
+bool Ekf::shouldInhibitMag() const
+{
+	// If the user has selected auto protection against indoor magnetic field errors, only use the magnetometer
+	// if a yaw angle relative to true North is required for navigation. If no GPS or other earth frame aiding
+	// is available, assume that we are operating indoors and the magnetometer should not be used.
+	// Also inhibit mag fusion when a strong magnetic field interference is detected
+	const bool user_selected = (_params.mag_fusion_type == MAG_FUSE_TYPE_INDOOR);
+	const bool not_using_gps = !(_params.fusion_mode & MASK_USE_GPS) || !_control_status.flags.gps;
+	const bool not_using_evpos = !(_params.fusion_mode & MASK_USE_EVPOS) || !_control_status.flags.ev_pos;
+	const bool not_using_evvel = !(_params.fusion_mode & MASK_USE_EVVEL) || !_control_status.flags.ev_vel;
+	const bool not_selected_evyaw =  !(_params.fusion_mode & MASK_USE_EVYAW);
+	const bool heading_not_required_for_navigation = not_using_gps &&
+							 not_using_evpos &&
+							 not_using_evvel &&
+							 not_selected_evyaw;
+
+	return (user_selected && heading_not_required_for_navigation)
+	       || isStrongMagneticDisturbance();
+}
+
+bool Ekf::isStrongMagneticDisturbance() const
+{
+	return _mag_sample_delayed.mag.length() > 2.f * _mag_strength_gps;
```

Comment:
That's a good question. I think it's really rare but it might be worth adding for completeness. I'll add a check for `_mag_sample_delayed.mag.length() < .5f * _mag_strength_gps`